pkg/stub/status.go

```diff
@@ -0,0 +1,200 @@
+	ns, err := h.manifestFactory.RouterNamespace()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"error building router namespace: %v", err)
+	}
+
+	if err := sdk.Get(ns); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil, nil, nil
+		}
+
+		return nil, nil, nil, fmt.Errorf(
+			"error getting Namespace %s: %v", ns.Name, err)
+	}
+
+	ingressList := &ingressv1alpha1.ClusterIngressList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterIngress",
+			APIVersion: "ingress.openshift.io/v1alpha1",
+		},
+	}
+	err = sdk.List(h.namespace, ingressList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list ClusterIngresses: %v", err)
+	}
+
+	daemonsetList := &appsv1.DaemonSetList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: "extensions/v1beta1",
+		},
+	}
+	err = sdk.List(h.namespace, daemonsetList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list DaemonSets: %v", err)
+	}
+
+	return ns, ingressList.Items, daemonsetList.Items, nil
+}
+
+// computeStatusConditions computes the operator's current state.
+func computeStatusConditions(conditions []osv1.ClusterOperatorStatusCondition, ns *corev1.Namespace, ingresses []ingressv1alpha1.ClusterIngress, daemonsets []appsv1.DaemonSet) []osv1.ClusterOperatorStatusCondition {
+	failingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorFailing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ns == nil {
+		failingCondition.Status = osv1.ConditionTrue
+		failingCondition.Reason = "NoNamespace"
+		failingCondition.Message = "router namespace does not exist"
+	} else {
+		failingCondition.Status = osv1.ConditionFalse
+	}
+	conditions = clusteroperator.SetStatusCondition(conditions,
+		failingCondition)
+
+	progressingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorProgressing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingresses != nil && daemonsets != nil {
+		numIngresses := len(ingresses)
+		numDaemonsets := len(daemonsets)
+		if numIngresses == numDaemonsets {
+			progressingCondition.Status = osv1.ConditionFalse
+		} else {
+			progressingCondition.Status = osv1.ConditionTrue
+			progressingCondition.Reason = "Reconciling"
+			progressingCondition.Message = fmt.Sprintf(
+				"have %d ingresses, want %d",
+				numDaemonsets, numIngresses)
+		}
+	}
+	conditions = clusteroperator.SetStatusCondition(conditions,
+		progressingCondition)
+
+	availableCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorAvailable,
+		Status: osv1.ConditionUnknown,
+	}
+	if ingresses != nil && daemonsets != nil {
+		dsAvailable := map[string]bool{}
+		for _, ds := range daemonsets {
+			dsAvailable[ds.Name] = ds.Status.NumberAvailable > 0
+		}
+
+		unavailable := []string{}
+		for _, ingress := range ingresses {
```

Comment:
Make a TODO here to revisit this block... the association logic needs cleaned up (e.g. using a function in `manifests` get associated resource names, or labels, or owner references, or something).

Reply:
Added.